### DISCUSSION TITLE
Renewing custom certs for Kafka cluster after they have expired

### DISCUSSION THREAD

**Discussion by vperi1730** [strimzi/strimzi-kafka-operator#5012] [Q&A]
Hi Scholzj,

Just 1 quick question, I have a Kafka cluster for which the custom-created certs have expired. I have followed the steps defined in the documentation https://strimzi.io/docs/0.17.0/#renewing-your-own-ca-certificates-str through which I was able to create new secrets that have the latest certs.

After this, I have first deleted the zookeeper pods and all the 3 of them re-created and running successfully. However, my Kafka pods are going into Crashloopbackoff. 

1) Am I missing something, Need your inputs here?
2) Will this procedure works even for expired certs?

```
kafka-2dayssl-entity-operator-7b696c954-skx86   3/3     Running            0          2d20h
kafka-2dayssl-kafka-0                           1/2     CrashLoopBackOff   5          4m53s
kafka-2dayssl-kafka-1                           1/2     CrashLoopBackOff   5          4m50s
kafka-2dayssl-kafka-2                           1/2     CrashLoopBackOff   5          4m43s
kafka-2dayssl-zookeeper-0                       2/2     Running            0          4m52s
kafka-2dayssl-zookeeper-1                       2/2     Running            0          4m43s
kafka-2dayssl-zookeeper-2                       2/2     Running            0          4m51s
strimzi-operator-6bb7c798d6-sql2r               1/1     Running            0          2d20h

```

These are the Kafka pod logs

```
2021-05-24 08:53:14,736 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-05-24 08:53:14,736 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-05-24 08:53:15,481 INFO [ZooKeeperClient Kafka server] Closing. (kafka.zookeeper.ZooKeeperClient) [main]
2021-05-24 08:53:16,508 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-05-24 08:53:16,509 INFO Socket connection established, initiating session, client: /127.0.0.1:38146, server: localhost/127.0.0.1:2181 (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-05-24 08:53:16,621 INFO Session: 0x0 closed (org.apache.zookeeper.ZooKeeper) [main]
2021-05-24 08:53:16,622 INFO EventThread shut down for session: 0x0 (org.apache.zookeeper.ClientCnxn) [main-EventThread]
2021-05-24 08:53:16,623 INFO [ZooKeeperClient Kafka server] Closed. (kafka.zookeeper.ZooKeeperClient) [main]
2021-05-24 08:53:16,626 ERROR Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer) [main]
kafka.zookeeper.ZooKeeperClientTimeoutException: Timed out waiting for connection while in state: CONNECTING
        at kafka.zookeeper.ZooKeeperClient.$anonfun$waitUntilConnected$3(ZooKeeperClient.scala:259)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:253)
        at kafka.zookeeper.ZooKeeperClient.waitUntilConnected(ZooKeeperClient.scala:255)
        at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:113)
        at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1858)
        at kafka.server.KafkaServer.createZkClient$1(KafkaServer.scala:375)
        at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:399)
        at kafka.server.KafkaServer.startup(KafkaServer.scala:207)
        at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:44)
        at kafka.Kafka$.main(Kafka.scala:84)
        at kafka.Kafka.main(Kafka.scala)
2021-05-24 08:53:16,627 INFO shutting down (kafka.server.KafkaServer) [main]
2021-05-24 08:53:16,633 INFO shut down completed (kafka.server.KafkaServer) [main]
2021-05-24 08:53:16,634 ERROR Exiting Kafka. (kafka.server.KafkaServerStartable) [main]
2021-05-24 08:53:16,636 INFO shutting down (kafka.server.KafkaServer) [kafka-shutdown-hook]
Heap
 garbage-first heap   total 8388608K, used 81920K [0x00000005c0000000, 0x00000005c0404000, 0x00000007c0000000)
  region size 4096K, 21 young (86016K), 0 survivors (0K)
 Metaspace       used 18106K, capacity 18712K, committed 18944K, reserved 1064960K
  class space    used 2351K, capacity 2464K, committed 2560K, reserved 1048576K
```

**Comment by scholzj** [comment]
@tombentley Is working on some fixes and improvements to the CA renewal. I think there were some bugs etc. But not sure what was the state of things in 0.17 which is quite old.

**Comment by vperi1730** [comment]
Ok, Is there a possibility @tombentley can throw some light here so it can help me resolve the issue? 

Just noticed something, after the new secrets are created tls-sidecar container within Kafka is throwing some strange errors that we are not sure about. Do you find any clue here?

```
Service [zookeeper-2181] accepted connection from 127.0.0.1:35650
2021.05.24 11:02:57 LOG5[1:140361699006208]: connect_blocking: connected 240.224.240.77:2181
2021.05.24 11:02:57 LOG5[1:140361699006208]: Service [zookeeper-2181] connected remote server from 240.0.4.167:58178
2021.05.24 11:02:57 LOG5[1:140361699006208]: Certificate accepted: depth=1, /CN=kafka.deloitte.com
2021.05.24 11:02:57 LOG4[1:140361699006208]: CERT: Verification error: certificate signature failure
2021.05.24 11:02:57 LOG4[1:140361699006208]: Certificate check failed: depth=0, /O=io.strimzi/CN=kafka-3dayssl-zookeeper
2021.05.24 11:02:57 LOG3[1:140361699006208]: error queue: 14090086: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
2021.05.24 11:02:57 LOG3[1:140361699006208]: error queue: D0C5006: error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib
2021.05.24 11:02:57 LOG3[1:140361699006208]: error queue: 4067072: error:04067072:rsa routines:RSA_EAY_PUBLIC_DECRYPT:padding check failed
2021.05.24 11:02:57 LOG3[1:140361699006208]: SSL_connect: 407008A: error:0407008A:rsa routines:RSA_padding_check_PKCS1_type_1:invalid padding
2021.05.24 11:02:57 LOG5[1:140361699006208]: Connection reset: 0 byte(s) sent to SSL, 0 byte(s) sent to socket
```

**Comment by vperi1730** [comment]
We have compared the ca-crt of the secret(<clustername>-cluster-ca-crt) with the one that tls-sidecar is using, both are identical. Even in this case, Kafka pods are still in Crashloopbackoff. 

Is there a way we can recover the cluster back in such situations where we manually renewed the certs and manually restarted the zk & kafka

**Comment by Gulnaz2021** [comment]
We've performed below mentioned scenarios:
**Scenario 1:**
We've launched a new kafka cluster (operator version 17 and application version 2.4.0) with manually created certificates, and the cluster is up and running.
Kafka spec:
  spec:
    clusterCa:
      generateCertificateAuthority: false
    clientsCa:
      generateCertificateAuthority: false
 
And the tls-sidecar logs are:
2021.05.25 07:33:10 LOG5[1:139884732647488]: stunnel 4.56 on x86_64-redhat-linux-gnu platform
2021.05.25 07:33:10 LOG5[1:139884732647488]: Compiled/running with OpenSSL 1.0.1e-fips 11 Feb 2013
2021.05.25 07:33:10 LOG5[1:139884732647488]: Threading:PTHREAD Sockets:POLL,IPv6 SSL:ENGINE,OCSP,FIPS Auth:LIBWRAP
2021.05.25 07:33:10 LOG5[1:139884732647488]: Reading configuration from file /tmp/stunnel.conf
2021.05.25 07:33:10 LOG5[1:139884732647488]: FIPS mode is enabled
2021.05.25 07:33:10 LOG4[1:139884732647488]: Insecure file permissions on /etc/tls-sidecar/kafka-brokers/kafka-3dayssl-kafka-0.key
2021.05.25 07:33:10 LOG5[1:139884732647488]: Configuration successful
2021.05.25 07:33:19 LOG5[1:139884732643072]: Service [zookeeper-2181] accepted connection from 127.0.0.1:49284
2021.05.25 07:33:19 LOG5[1:139884732643072]: connect_blocking: connected 172.24.0.188:2181
**2021.05.25 07:33:19 LOG5[1:139884732643072]: Service [zookeeper-2181] connected remote server from 240.0.24.242:60500
2021.05.25 07:33:19 LOG5[1:139884732643072]: Certificate accepted: depth=1, /CN=kafka.xyz.com
2021.05.25 07:33:19 LOG5[1:139884732643072]: Certificate accepted: depth=0, /O=io.strimzi/CN=kafka-3dayssl-zookeeper
2021.05.25 07:33:24 LOG5[1:139884732573440]: Service [zookeeper-2181] accepted connection from 127.0.0.1:49516**
2021.05.25 07:33:24 LOG5[1:139884732573440]: connect_blocking: connected 172.24.0.188:2181
2021.05.25 07:33:24 LOG5[1:139884732573440]: Service [zookeeper-2181] connected remote server from 240.0.24.242:60730

**Scenario 2:**
We've renewed the certificates on the same cluster and we did a rolling restart on zookeeper, and all of zookeeper pods restarted and are running. Now, upon doing the same for Kafka pods, they went to crashloopbackoff. And we've recreated all the mentioned below 4 secrets:
<cluster name>-clients-ca
<cluster name>-clients-ca-cert
<cluster name>-cluster-ca
<cluster name>-cluster-ca-cert
And when we checked the tls-sidecar logs, mentioned below:
**2021.05.24 08:25:05 LOG5[1:140343124711168]: Service [zookeeper-2181] connected remote server from 240.0.6.246:35230
2021.05.24 08:25:05 LOG5[1:140343124711168]: Certificate accepted: depth=1, /CN=kafka.xyz.com
2021.05.24 08:25:05 LOG4[1:140343124711168]: CERT: Verification error: certificate signature failure
2021.05.24 08:25:05 LOG4[1:140343124711168]: Certificate check failed: depth=0, /O=io.strimzi/CN=kafka-2dayssl-zookeeper**
2021.05.24 08:25:05 LOG3[1:140343124711168]: error queue: 14090086: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed

Observed that certs with depth=1 has been accepted and certs with depth=0 has failed.
The only difference in scenario 1 and 2 is that, we've renewed the certificates.
Hence, wanted to know why have we observed this gap. Can you please let me know, a resolution to this.

**Reply by scholzj** [reply]
Is the custom CA you use configure as a CA? Does it have the CA:true flag (or constraint or whatever it is called in the X509 terminology)?

**Reply by scholzj** [reply]
You can also check if the actual server certificates in the ZooKeeper and Kafka secrets were updated. That could probably cause similar errors.

**Reply by vperi1730** [reply]
> CA:true

Yes I have verified the cert inside the secret and it has CA: true flag

```
 X509v3 extensions:
            X509v3 Subject Key Identifier:
                9A:4B:26:8B:DD:2A:DB:E3:87:3D:E6:59:5C:D8:ED:77:E0:AC:17:06
            X509v3 Authority Key Identifier:
                keyid:9A:4B:26:8B:DD:2A:DB:E3:87:3D:E6:59:5C:D8:ED:77:E0:AC:17:06

            X509v3 Basic Constraints:
                CA:TRUE
    Signature Algorithm: sha256WithRSAEncryption

```
You can also check if the actual server certificates in the ZooKeeper and Kafka secrets were updated. That could probably cause similar errors. - For this one 4 Kafka secrets are something that we have created as part of the renewal, Do you want us to verify if the secret cluster name-zookeeper-nodes that contain all the 3 nodes' certs are identical to what we have in Kafka. Is our understanding correct?

**Reply by scholzj** [reply]
> For this one 4 Kafka secrets are something that we have created as part of the renewal, Do you want us to verify if the secret cluster name-zookeeper-nodes that contain all the 3 nodes' certs are identical to what we have in Kafka. Is our understanding correct?

Well, the secrets you listed above are just the CAs. They are not the server certs. They will be stored in the `-zookeeper-nodes` and `-kafka-brokers` secrets. So you should check if those were updated and if not, delete them to have them recreated.

**Reply by Gulnaz2021** [reply]
As suggested above, we've compared the below mentioned .crt values of the -zookeeper-nodes with the ca.crt value of the renewed certificate kafka-2dayssl-cluster-ca-cert:
kafka-2dayssl-zookeeper-0.crt,
kafka-2dayssl-zookeeper-1.crt,
kafka-2dayssl-zookeeper-2.crt
Observed that none of them are same. 

Similarly, we've compared the below mentioned .crt values of the -kafka-brokers with the ca.crt value of the renewed certificate kafka-2dayssl-cluster-ca-cert:
kafka-2dayssl-kafka-0.crt,
kafka-2dayssl-kafka-1.crt,
kafka-2dayssl-kafka-2.crt
Observed that none of them are same. 

Kafka spec:
spec:
clusterCa:
generateCertificateAuthority: false
clientsCa:
generateCertificateAuthority: false

Hence, we've deleted the secrets -zookeeper-nodes and -kafka-brokers, but observed that -zookeeper-nodes got recreated, but -kafka-brokers didn't recreate. Please let us know, how much time does it takes to recreate the secret or  do we have to do it some other way.

**Reply by scholzj** [reply]
Sorry, but WDYM with `Observed that none of them are same.`? They are not going to be the same. The question is whether they are signed by your renewed CA (actually, did you renewed or replaced the CA? Is the private key still the same or not?).

> Hence, we've deleted the secrets -zookeeper-nodes and -kafka-brokers, but observed that -zookeeper-nodes got recreated, but -kafka-brokers didn't recreate. Please let us know, how much time does it takes to recreate the secret or do we have to do it some other way.

AFAIK in the current version, the `kafka-brokers` secret will be recreated after the Zookeeper is done. So if it was not I guess that ZooKeeper is still not working? But it could be that this was different in 0.17 ... a lot changed since.

**Reply by vperi1730** [reply]
1) We have followed this section to renew it(https://strimzi.io/docs/operators/in-development/using.html#renewing-your-own-ca-certificates-str), however, the procedure says that we need to delete the secret and re-create them using OpenSSL mentioning the no of days. 

2) We have created a new private key also. This is the script that we executed

```
#!/bin/sh

NO_OF_DAYS=10
COMMON_NAME="kafka.deloitte.com"
NAMESPACE="do-kafka-ssluc1"
CLUSTERNAME="kafka-2dayssl"

#########################Generate Cluster Certificates###########################

openssl req -new -newkey rsa:4096 -days $NO_OF_DAYS -x509 -subj "/CN=$COMMON_NAME" -keyout $CLUSTERNAME-server.key -out $CLUSTERNAME-server.crt -nodes

kubectl -n $NAMESPACE create secret generic ${CLUSTERNAME}-cluster-ca-cert --from-file=ca.crt=$CLUSTERNAME-server.crt

kubectl -n $NAMESPACE label secret ${CLUSTERNAME}-cluster-ca-cert strimzi.io/kind=Kafka strimzi.io/cluster=$CLUSTERNAME

kubectl -n $NAMESPACE create secret generic ${CLUSTERNAME}-cluster-ca --from-file=ca.key=$CLUSTERNAME-server.key

kubectl -n $NAMESPACE label secret ${CLUSTERNAME}-cluster-ca strimzi.io/kind=Kafka strimzi.io/cluster=$CLUSTERNAME

###############Generate Clients Certificates######################################

openssl req -new -newkey rsa:4096 -days $NO_OF_DAYS -x509 -subj "/CN=$COMMON_NAME" -keyout $CLUSTERNAME-client.key -out $CLUSTERNAME-client.crt -nodes

kubectl -n $NAMESPACE create secret generic ${CLUSTERNAME}-clients-ca-cert --from-file=ca.crt=$CLUSTERNAME-client.crt

kubectl -n $NAMESPACE label secret ${CLUSTERNAME}-clients-ca-cert strimzi.io/kind=Kafka strimzi.io/cluster=$CLUSTERNAME

kubectl -n $NAMESPACE create secret generic ${CLUSTERNAME}-clients-ca --from-file=ca.key=$CLUSTERNAME-client.key

kubectl -n $NAMESPACE label secret ${CLUSTERNAME}-clients-ca strimzi.io/kind=Kafka strimzi.io/cluster=$CLUSTERNAME
```

**Reply by scholzj** [reply]
Ok, I did not know about this docs ... but that is replacing the private key. So all the certificates signed by the previous CA will not be valid anymore and will need to be updated. Deleting the secrets should normally force them to be updated (and might be needed until the fixes @tombentley is working on).

**Comment by Gulnaz2021** [comment]
We were able to renew the certificates of Kafka cluster by following the steps provided in the document https://strimzi.io/docs/0.17.0/#renewing-your-own-ca-certificates-str. And post this we've performed rolling restart of the pods in the cluster and cluster has been deployed successfully. Then we were trying to produce and consume messages using the truststore and keystore. 
As a part of renewal, secrets <cluster>-clients-ca-cert and <cluster>-clients-ca-cert include the below mentioned two .crt files,
ca-2021-05-29T06-22-26Z.crt
ca.crt
Now when we are producing and consuming messages using truststore(generated with the old ca.crt and the renewed ca.crt) and keystore, we are getting below mentioned error:
**Producer:**
ERROR [Producer clientId=console-producer] Connection to node -1 (kafka-cluster-kafka-bootstrap/xxx.xxx.xx.xx:9093) failed authentication due to: SSL handshake failed (org.apache.kafka.clients.NetworkClient)

**Consumer:**
ERROR Error processing message, terminating consumer process:  (kafka.tools.ConsoleConsumer$)
org.apache.kafka.common.errors.SslAuthenticationException: SSL handshake failed
Caused by: javax.net.ssl.SSLHandshakeException: General SSLEngine problem
        at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1521)
        at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:528)
        at sun.security.ssl.SSLEngineImpl.writeAppRecord(SSLEngineImpl.java:1197)
        at sun.security.ssl.SSLEngineImpl.wrap(SSLEngineImpl.java:1165)
        at javax.net.ssl.SSLEngine.wrap(SSLEngine.java:469)
        at org.apache.kafka.common.network.SslTransportLayer.handshakeWrap(SslTransportLayer.java:448)
        at org.apache.kafka.common.network.SslTransportLayer.doHandshake(SslTransportLayer.java:313)
        at org.apache.kafka.common.network.SslTransportLayer.handshake(SslTransportLayer.java:265)
Caused by: sun.security.validator.ValidatorException: PKIX path validation failed: java.security.cert.CertPathValidatorException: Path does not chain with any of the trust anchors
        at sun.security.validator.PKIXValidator.doValidate(PKIXValidator.java:380)
        at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:285)
        at sun.security.validator.Validator.validate(Validator.java:262)
        at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:330)
        at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:289)
        at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:144)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1626)
        ... 27 more

Can I get a resolution for this.

**Reply by scholzj** [reply]
`PKIX path validation failed` suggests that the truststore does not match the server. To check more, you can set the Java system property `-Djavax.debug.net=ssl` (for example using the `KAFKA_OPTS` env var.). That should show more what do you have in the truststore and what the broker is presenting.

**Reply by Gulnaz2021** [reply]
During the renewal of certificates, why are we extracting the old ca-2021-05-29T06-22-26Z.crt along with the new ca.crt and using it in the creation of (<cluster>-cluster-ca or <cluster>-clients-ca) and  (<cluster>-cluster-ca-cert or <cluster>-clients-ca-cert). 
As when I've added both the certs to my truststore, I was able to consume and produce messages. Hence, can you help me with below questions as well,
```
Once the old certificates (ca-2021-05-29T06-22-26Z.crt ) would expire (in my case it is expiring tomorrow), 
1) would it still be still valid, 
2) how would it impact my truststore, would I still be able to produce and consume,
3) how would it impact the cluster,
4) how would it impact the pod,
```

**Reply by scholzj** [reply]
Sorry, I do not follow this. Is it part of the documentation? Or why did you do it? In general, the old public key in the truststore would be needed during the transition as it would allow the clients to connect to the brokers still using the old certificates. After it expires, it should not be possible to use it AFAIK, so you can just delete it.

**Comment by Gulnaz2021** [comment]
1) I've deployed a kafka cluster with custom certs and the expiration of the certs was 2 days. And I was able to produce and consume messages with the truststore and keystore.

2) Now, I've manually renewed the certificates of Kafka cluster by following the steps provided in the document https://strimzi.io/docs/0.17.0/#renewing-your-own-ca-certificates-str. And have mentioned the expiration of certs for 10 days. And post this I've performed rolling restart of the pods in the cluster and cluster has been deployed successfully. And I was able to produce and consume messages using the truststore and keystore. 
As a part of renewal of certificates, creation of secrets (-cluster-ca or -clients-ca) and (-cluster-ca-cert or -clients-ca-cert) include the below mentioned two .crt files,
(old) ca-2021-05-29T06-22-26Z.crt
(new) ca.crt
and have added both the ca.crt and ca-2021-05-29T06-22-26Z.crt to the truststore.

3) I've followed the similar approach as mentioned in step 2, to again renew the certificates(that have already been renewed once in step2) of kafka cluster with an expiration of 10 days. So, am renewing the certificate twice now. 
Creation of the secrets (-cluster-ca or -clients-ca) and (-cluster-ca-cert or -clients-ca-cert) include the below mentioned two .crt files,
ca-2021-06-06T10-21-49Z.crt (old ca.crt that is extracted from kafka-cluster-cluster-ca-cert generated in step -2) 
ca.crt (new)
Post this while performing the rolling restart of the pods in the cluster, kafka pods are going to crashloopbackoff status, with the tls-sidecar logs as:

021.06.03 11:33:24 LOG5[1:140491840714496]: Service [zookeeper-2181] connected remote server from 240.0.22.179:49322
2021.06.03 11:33:24 LOG5[1:140491840714496]: Certificate accepted: depth=1, /CN=xyz.com
2021.06.03 11:33:24 LOG4[1:140491840714496]: CERT: Verification error: certificate signature failure
2021.06.03 11:33:24 LOG4[1:140491840714496]: Certificate check failed: depth=0, /O=io.strimzi/CN=kafka-cluster-zookeeper
2021.06.03 11:33:24 LOG3[1:140491840714496]: error queue: 14090086: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
2021.06.03 11:33:24 LOG3[1:140491840714496]: error queue: D0C5006: error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib
2021.06.03 11:33:24 LOG3[1:140491840714496]: error queue: 4067072: error:04067072:rsa routines:RSA_EAY_PUBLIC_DECRYPT:padding check failed
2021.06.03 11:33:24 LOG3[1:140491840714496]: SSL_connect: 407008A: error:0407008A:rsa routines:RSA_padding_check_PKCS1_type_1:invalid padding
2021.06.03 11:33:24 LOG5[1:140491840714496]: Connection reset: 0 byte(s) sent to SSL, 0 byte(s) sent to socket

Can I get a resolution for this error.

**Reply by scholzj** [reply]
I guess the ZooKeeper certificates did not renew? You should probably check the other secrets if they were updated before you roll the pods. Because if you update the CA secrets but the operator doesn't update the server certificates, rolling the pods will not work.

**Reply by Gulnaz2021** [reply]
With zookeeper certificates, do you mean secret - <>-zookeeper-nodes, as this is same, since the duration cluster has been deployed with custom certificates and
with the same secret <>-zookeeper-nodes, the first manual renewal of the certificate worked.
But when we are trying to renew the already renewed certificate - that is mentioned in step 3  above, observing errors. Can you please help with the information that :
1) whether it is possible with operator version 17, manual renewal of the secrets twice, or there is some other way that needs to be tried out.
2) suppose I have 3 ca.crts, where first 2 are appended with the time stamp and the third one is latest generated ca.crt, and 1st ca.crt is expired. Then, if I'll chain all the three ca.crts to generate cluster/clients-ca-cert, then will the rolling restart work in this case and whether the kafka cluster would be in the running state.
Example:
ca-2021-06-10T08-44-50Z.crt - first expired certificate
ca-2021-06-13T11-35-30Z.crt - second active certificate
ca.crt - latest active certificate
3) after the expiration of a certificate, what is the procedure to bring back the kafka cluster to the running state.

**Reply by scholzj** [reply]
Yes, correct. With ZooKeeper certificates I mean the certs int he `<cluster-name>-zookeeper-nodes`. These are the server certificates the ZooKeeper node use. If the ZooKeeper certs didn't renew, then the step two will still work as you still include the old `ca.crt` (as `ca-2021-05-29T06-22-26Z.crt`). But if you remove it in the step 3, they will nto be trusted anymore and things break. So you need to renew the ZooKeeper certs first ... one way to do that might be to delete the secret and have it recreated. (it should normally happen automatically, not not sure if there were any bugs in 0.17 or not)

**Reply by Gulnaz2021** [reply]
For operator version 17 with application - 2.4.0 and version 19 with application - 2.5.0, provided first secrets is expired, and with the second secret that is active, am renewing the second secret again; and my secrets - 
kafka-cluster-clients-ca-cert - has the ca.crt and the ca.crt appended with the date of the second secret
kafka-cluster-cluster-ca-cert- has the ca.crt and the ca.crt appended with the date of the second secret.

I've compared the below mentioned certs in the secret ``<cluster-name>-zookeeper-nodes``, prior and post manual renewal of the certificates,
kafka-cluster-zookeeper-0.crt
kafka-cluster-zookeeper-1.crt
kafka-cluster-zookeeper-2.crt
But have observed that, the certs in the secret ``<cluster-name>-zookeeper-nodes``, were same.
As suggested, I've deleted the secret ``<cluster-name>-zookeeper-nodes``, and it got recreated. But, despite of that upon rolling restart of the pods,
** 1) with version 17 operator:**
Observed that zookeeper pods were in running state, but for the kafka pods went to crashloopbackoff state with the error:

kubectl -n <ns> logs kafka-cluster-kafka-0  -c tls-sidecar

 Connection reset: 0 byte(s) sent to SSL, 0 byte(s) sent to socket
2021.06.14 11:27:12 LOG5[1:140243778553600]: Service [zookeeper-2181] accepted connection from 127.0.0.1:39266
2021.06.14 11:27:12 LOG5[1:140243778553600]: connect_blocking: connected 240.224.161.70:2181
2021.06.14 11:27:12 LOG5[1:140243778553600]: Service [zookeeper-2181] connected remote server from 240.0.12.141:54834
2021.06.14 11:27:12 LOG5[1:140243778553600]: Certificate accepted: depth=1, /CN=xyz.com
2021.06.14 11:27:12 LOG5[1:140243778553600]: Certificate accepted: depth=0, /O=io.strimzi/CN=kafka-cluster-zookeeper
2021.06.14 11:27:12 LOG3[1:140243778553600]: SSL_connect: 1409441B: error:1409441B:SSL routines:ssl3_read_bytes:tlsv1 alert decrypt error
2021.06.14 11:27:12 LOG5[1:140243778553600]: Connection reset: 0 byte(s) sent to SSL, 0 byte(s) sent to socket
2021.06.14 11:27:14 LOG5[1:140243778553600]: Service [zookeeper-2181] accepted connection from 127.0.0.1:39320
2021.06.14 11:27:14 LOG5[1:140243778553600]: connect_blocking: connected 240.224.161.70:2181
2021.06.14 11:27:14 LOG5[1:140243778553600]: Service [zookeeper-2181] connected remote server from 240.0.12.141:54888
2021.06.14 11:27:14 LOG5[1:140243778553600]: Certificate accepted: depth=1, /CN=xyz.com
2021.06.14 11:27:14 LOG5[1:140243778553600]: Certificate accepted: depth=0, /O=io.strimzi/CN=kafka-cluster-zookeeper
2021.06.14 11:27:14 LOG3[1:140243778553600]: SSL_connect: 1409441B: error:1409441B:SSL routines:ssl3_read_bytes:tlsv1 alert decrypt error

** 2) with version 19 operator:**
Observed that zookeeper pods were in running state, but for the kafka pods went to crashloopbackoff state with the error:

kubectl -n <ns> kafka-cluster-kafka-0 -c tls-sidecar -f

2021.06.14 11:18:41 LOG5[19:139658516170816]: Reading configuration from file /tmp/stunnel.conf
2021.06.14 11:18:41 LOG5[19:139658516170816]: FIPS mode is enabled
2021.06.14 11:18:41 LOG4[19:139658516170816]: Insecure file permissions on /etc/tls-sidecar/kafka-brokers/kafka-cluster-kafka-0.key
2021.06.14 11:18:41 LOG5[19:139658516170816]: Configuration successful

Let me know if there is any bug with version 17 as stated above and why are the kafka pods going to crashloopbackoff  with version 19 as well. Or is there any other possible solution to it?

**Reply by scholzj** [reply]
Well, your log from 0.19 does not show any errors. So I do not know why it is crashlooping.

**Reply by Gulnaz2021** [reply]
With version 0.17 operator,
As we've followed the below steps:
1) Renewing the certificates manually,
2) Deleted the secret ``<cluster-name>-zookeeper-nodes``, until it is recreated,
3) Rolling restart of the kafka pods, starting from zookeeper to kafka, 
Observed that kafka pods are going to crashloopbackoff with the above error for tls-sidecar and with the below logs for kafka container:

kubectl -n ns logs kafka-cluster-kafka-0 -c kafka,

2021-06-14 12:51:04,840 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-06-14 12:51:04,841 INFO Socket connection established, initiating session, client: /127.0.0.1:34876, server: localhost/127.0.0.1:2181 (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2021-06-14 12:51:04,845 WARN Session 0x0 for server localhost/127.0.0.1:2181, unexpected error, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
java.io.IOException: Connection reset by peer
        at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
        at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
        at sun.nio.ch.IOUtil.read(IOUtil.java:192)
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:377)
        at org.apache.zookeeper.ClientCnxnSocketNIO.doIO(ClientCnxnSocketNIO.java:75)
        at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:363)
        at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1223)

Is it that we are facing this issue because the first ca.crt has been expired. Can you please update why it is not working with version 0.17, or is there any bug and in which version it has been further resolved.

**Reply by scholzj** [reply]
Well, as I said on the beginning ... there are known issues to the custom CA renewal which @tombentley is working on. But 0.17 (There is no version 17, it is 0.17) or 0.19 is long time ago, I do not know anymore if it had or had not some issues fixed later, sorry. But it should be all tracked in the PRs / issues. So you can check them.
